_posts/2021-02-10-hindsight-on-2020.md

```diff
@@ -0,0 +1,254 @@
+with flying colors. It also gained the ability to encode indexed data directly.
+
+The `tiff` crate has also seen a lot of attention. Throughout the year various
+contributors added BigTIFF and basic GeoTIFF support, have expanded the
+supported sample formats to 32-bit and 64-bit color depth, and overall
+increased the amount of recognized tags. It also decodes embedded modern jpeg
+image data.  Nevertheless, one particularly sore point remains: `tiff` does not
+yet support Tile Offsets which would be extremely valuable for access to large
+data from medical and geodetic data. We'd be delighted to see contributions for
+this feature.
+
+The `jpeg-decoder` crate now correctly decodes all progressive images. It now
+collects ICC and Exif data and makes them available in their raw form. The
+implementation exposing these in a general interface within `image` is expected
+to commence sometime soon.
+
+We've tested the new implementations on web data dumps of hundreds of thousands
+of images from the web—totalling terabytes of data— and decoded the
+overwhelming majority of them successfully.
+
+### Performance
+
+The `gif` crate joins `png` in being on par with the relevant C implementations
+in terms of performance! This was made possible by a combination of
+improvements. Firstly, the restructured decoder interface make it possible to
+write directly into the output buffer which avoids a few copies. Secondly, the
+decompression libraries have been upgraded.
+
+These gains are thanks to a new LZW encoding/decoding library, [called
+`weezl`][weezl]. It significantly improves on lzw crate in performance while
+staying 100% safe, and also had an optional `no_std` mode. There is integration
+with both `std` and `async` IO-interfaces, behind separate features. `weezl`,
+written by the author of this post, is also maintained under the image-rs
+organization.
+
+Since most of the time in decoding GIF files is spent in LZW decoding, the
+gains from LZW have naturally translated to gif crate. TIFF files encoded with
+LZW compression also benefit from this change. The total decoding throughput
+quadrupled in some instances! The latest versions of gif, tiff and image all
+take advantage of this new library.
+
+[weezl]: https://crates.io/crates/weezl
+
+The `jpeg-decoder` crate has also been sped up significantly. Idiomatic
+iterator based loops of Rust allow the LLVM compiler to aggressively use auto
+vectorization. That is, instead of performing a transformation on single
+elements at a time this optimization technique automatically replaces code
+operating on elements sequentially (in this case, pixels) with special SIMD
+instructions that perform them on multiple elements at a time, with large
+speedups of 400% or more. While not yet on-par with handwritten assembly this
+can get very close with the right code structure. Additionally, a number of
+index checks have been identified. Together, the whole decoding process takes
```

Comment:
> Additionally, a number of index checks have been identified.

Not super clear. Try: "Iterator-based approach also eliminated some needless bounds checks, which improved performance without sacrificing safety."

Reply:
Also, you forgot to include increased parallelism - IDCT is now threaded per-component rather than a single worker thread: https://github.com/image-rs/jpeg-decoder/pull/168